### PR TITLE
don't assume a file uri is absolute

### PR DESCRIPTION
**Pull request by brettfo**
_Fixes internal bug 165369._

We received customer feedback about a crash in VS that we weren't able to repro but contained a dump.  Examining the dump revealed that the error was the `AbsolutePath` property on `System.Uri` was being accessed on a relative (non-absolute) `Uri`.  Inspecting all `System.Uri` heap objects reported a number of items with relative paths which would have caused the exception seen in the dump.  See below for example `Uri`s extracted from the dump.

The fix is to instead use the `OriginalString` property on `System.Uri`, because that's the only property that's guaranteed not to throw on a relative `Uri`.  I've also confirmed that [`Path.GetExtension()`](http://referencesource.microsoft.com/#mscorlib/system/io/path.cs,f424e433705aeb09) will only ever return `null` if its parameter (`OriginalString`) is `null`, but the [`System.Uri` constructor](http://referencesource.microsoft.com/#System/net/System/URI.cs,d9fc45d783ee73ff,references) explicitly disallows this, so the `null` check was able to be removed.

**Example `Uri`s:**
- `file:///C:/WebApplication1/WebApplication1/WebApplication1.vbproj` // absolute
- `/C:/WebApplication1/WebApplication1/Startup.vb` // relative, and likely the cause of the crash

**Follow up:**
Contact other internal teams to inform them that they're mixing relative and absolute `Uri`s in the project system.

@dotnet/roslyn-ide

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
I find your lack of tests disturbing.

**Comment by brettfo** [comment]
It's not for lack of trying.  To hit this specific code path [`graphBuilder.GetContextDocument(node)`](https://github.com/dotnet/roslyn/pull/7996/files#diff-d6d8c938bacc74451ff7f20168c775fdR33) has to return `null` and I can't for the life of me see how we could get a graph with a node that's not either a symbol or a document (which hits other branches in this function).  I'm certainly open to suggestions, but I've tried adding non-code files to the test project, links, etc., but I just can't hit this path.

**Comment by dpoeschl** [comment]
@brettfo It seems like you just need `GraphBuilder.CreateForInputNodesAsync` to return a custom, test-only `GraphBuilder` (or newly introduced `IGraphBuilder`) that returns null for the symbol and document, right? If so, you could remove the static dependency on `GraphBuilder.CreateForInputNodesAsync` method that these `IGraphQuery` implementations use and instead pass in some sort of `IGraphBuilderFactory` that can create real ones in production code or contrived ones in test code. If that's insufficient, what else am I missing?

**Comment by jasonmalinowski** [comment]
`GraphBuilder` is our own type, we don't need to mock it or anything. @brettfo: it could happen as long as somebody created a node and queried against us, but we ourselves didn't create that node. Fundamentally, I made an architectural mistake when I wrote that code where it more or less assumes that all the nodes we answer stuff for _must_ be nodes we created ourselves, at least for symbols. The old provider didn't require that to the best of my understanding.

**Comment by brettfo** [comment]
Thanks @dpoeschl and @jasonmalinowski for the ideas.  I was over-thinking it and instead manually created a graph node similar to one I found in the dump.  I confirmed that without my fix the test fails with the same exception found in the dump.

**Comment by brettfo** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/2880/
// Retest reason: build timeout

**Comment by brettfo** [comment]
Ping for additional feedback/signoff, especially since a test has been added.

File: src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/ContainsChildrenGraphQuery.cs [hunk 1]
```diff
@@ -53,9 +53,11 @@ internal sealed class ContainsChildrenGraphQuery : IGraphQuery
                                 // even when its extension fails to say so. One option would be to call DTEWrapper.IsRegisteredForLangService,
                                 // which may not be called here however since deadlock could happen.
 
-                                string ext = Path.GetExtension(uri.AbsolutePath);
-                                if (!string.IsNullOrEmpty(ext) &&
-                                    (ext.Equals(".cs", StringComparison.OrdinalIgnoreCase) || ext.Equals(".vb", StringComparison.OrdinalIgnoreCase)))
+                                // The Uri returned by `GetNestedValueByName()` above isn't necessarily absolute and the `OriginalString` is
+                                // the only property that doesn't throw if the UriKind is relative, so `OriginalString` must be used instead
+                                // of `AbsolutePath`.
+                                string ext = Path.GetExtension(uri.OriginalString);
```
Comment (jmarolf) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/ContainsChildrenGraphQuery.cs]: To clarify `Path.GetExtension(uri.OriginalString)` is guaranteed to never return null but `Path.GetExtension(uri.AbsolutePath)` will return null sometimes? Or was the original code being paranoid checking if `ext` was null?

Comment (brettfo) [inline src/VisualStudio/Core/Def/Implementation/Progression/GraphQueries/ContainsChildrenGraphQuery.cs]: The original code was being paranoid; it's impossible to create a Uri with a null path and `Path.GetExtension()` never returns null, so just grabbing the extension and doing the comparison is sufficient.

**Comment by jmarolf** [comment]
:+1:

**Comment by dpoeschl** [comment]
:+1:

**Pull request merged**